### PR TITLE
feat(worker): Allow using a specific brigade-worker image per project

### PR DESCRIPTION
**Pull request by mumoshu**
`worker.(registry|name|tag|pullPolicy)` are added to brigade-project's values.yaml so that
you can differentiate which docker image to be used for running brigade.js per-project basis.

Ref #240

This is a minimum change seemingly required in order to support reusability of node modules, possibly hosted in private git repositories, plus arbitrary binaries usable from within brigade.js'.

### CODE REVIEW

**Comment by technosophos** [comment]
This can definitely help with the JavaScript library problem. And it's useful in general, since we do want to make it easy to write custom workers. So I'm definitely in favor of adding this. I'm going to give @adamreese a chance to review before I merge it though.

To get all of the JS support we want for #240, we might also need to remove the whitelist code for `require` in `index.ts`.

**Comment by mumoshu** [comment]
> To get all of the JS support we want for #240, we might also need to remove the whitelist code for require in index.ts.

Thanks for the info!
I've addressed it in the last commit.

File: brigade-worker/src/require.ts [hunk 1]
```diff
@@ -1,24 +1,6 @@
...
-  // library.
-  // {name: "crypto"},
-  // buffer.allocUnsafe could potentially allow other parts of the program to be
-  // accessed. This could presumably include something like a Kubernetes token.
-  // The present security model does not consider this to be a critical issue,
-  // since this information is obtainable via other pods as well. So we leave
-  // the buffer module. In the future, it might be good to mask that function.
-  { name: "buffer" },
-  { name: "assert" }
+  { name: "brigadier", override: "./brigadier" }
```
Comment (technosophos) [inline brigade-worker/src/require.ts]: Excellent. /cc @blimmer

**Comment by technosophos** [comment]
One of my recent merges introduced a merge conflict. Would you mind rebasing?

**Comment by mumoshu** [comment]
@technosophos Sure! Just finished rebasing. Would you mind taking a look again?

**Comment by mumoshu** [comment]
Hi @adamreese, this is just a friendly reminder that we're looking forward to your review. Thanks!

File: pkg/brigade/project_test.go [hunk 2]
```diff
@@ -43,3 +43,37 @@ func TestProjectSecrets(t *testing.T) {
 		t.Error("Project.Repo.SSHKey should not be exported")
 	}
 }
+
+func TestProjectWorkerConfig(t *testing.T) {
```
Comment (adamreese) [inline pkg/brigade/project_test.go]: What is this testing?

Comment (mumoshu) [reply pkg/brigade/project_test.go]: This is testing that `WorkerConfig` can be serialized to/deserialized from JSON.
But I'm open to suggestion like e.g. merging this test into [`TestProjectSecrets`](https://github.com/Azure/brigade/pull/277/files/c6cfb753a2ed9eb6a9e2e47dec549e14d5a742eb#diff-df2235e76b97ddcdf1c446d032d54964R19)!

File: charts/brigade-project/values.yaml [hunk 3]
```diff
@@ -102,3 +102,11 @@ kubernetes:
...
 # Uncomment the line below to increase logging verbosity.
 # workerCommand: "yarn start"
+
+# OPTIONAL: Project-specific worker settings which takes precedence over brigade-wide settings
+# Useful when you want a specific brigade-worker image for running brigade.js of this project.
+worker:
+  #registry: deis
+  #name: brigade-worker
+  #tag:
+  #pullPolicy: IfNotPresent
```
Comment (mumoshu) [inline charts/brigade-project/values.yaml]: I believe this `worker.*` keys should be consistent vs `workerCommand`.
I'll open a dedicated github issue for discussion if/once this gets merged.

Comment (mumoshu) [reply charts/brigade-project/values.yaml]: Also, we should have a general rule like "If two or more keys are prefixed by the same word, we should make them nested under the word".

**Pull request merged**